cpp/src/stream_compaction/drop_nans.cu

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/detail/copy_if.cuh>
+#include <cudf/detail/nvtx/ranges.hpp>
+#include <cudf/detail/stream_compaction.hpp>
+#include <cudf/stream_compaction.hpp>
+#include <cudf/table/table.hpp>
+#include <cudf/table/table_device_view.cuh>
+#include <cudf/table/table_view.hpp>
+#include <cudf/utilities/type_dispatcher.hpp>
+
+namespace {
+
+struct dispatch_is_not_nan {
+  template <typename T>
+  std::enable_if_t<std::is_floating_point<T>::value, bool> __device__
+  operator()(cudf::column_device_view col_device_view, cudf::size_type i)
+  {
+    return col_device_view.is_valid(i) ? not std::isnan(col_device_view.element<T>(i)) : true;
+  }
+
+  template <typename T>
+  std::enable_if_t<not std::is_floating_point<T>::value, bool> __device__
+  operator()(cudf::column_device_view col_device_view, cudf::size_type i)
+  {
+    return true;
+  }
+};
+
+// Returns true if the mask is true and it is not NaN for index i in at least keep_threshold
+// columns
+struct valid_table_filter {
+  __device__ inline bool operator()(cudf::size_type i)
+  {
+    auto valid = [i](auto col_device_view) {
+      return cudf::type_dispatcher(
+        col_device_view.type(), dispatch_is_not_nan{}, col_device_view, i);
```

Comment:
Shouldn't drop_nans be identical to dop_nulls with the only difference being to pass a different functor here? I was expecting this to be a much smaller PR.

Reply:
I was thinking same, but there are couple of checks before it can be called. 

I can make changes to drop_nulls with a Boolean parameter to drop nulls or NANs, but it would look odd, and naming would be clumsy.